torchvision/prototype/datapoints/_dataset_wrapper.py

```diff
@@ -0,0 +1,250 @@
+        self.transforms, dataset.transforms = dataset.transforms, None
+
+    def __getattr__(self, item: str) -> Any:
+        with contextlib.suppress(AttributeError):
+            return object.__getattribute__(self, item)
+
+        return getattr(self.vision_dataset, item)
+
+    def __getitem__(self, idx: int) -> Any:
+        # This gets us the raw sample since we disabled the transforms for the underlying dataset in the constructor
+        # of this class
+        sample = self.vision_dataset[idx]
+
+        sample = self.wrapper(self.vision_dataset, sample)
+
+        # We don't need to care about `transform` and `target_transform` here since `VisionDataset` joins them into a
+        # `transforms` internally:
+        # https://github.com/pytorch/vision/blob/2d92728341bbd3dc1e0f1e86c6a436049bbb3403/torchvision/datasets/vision.py#L52-L54
+        if self.transforms is not None:
+            sample = self.transforms(*sample)
+
+        return sample
+
+    def __len__(self) -> int:
+        return len(self.vision_dataset)
+
+
+@functools.lru_cache(maxsize=None)
+def get_categories(dataset: datasets.VisionDataset) -> Optional[List[str]]:
+    categories_fn = {
+        datasets.Caltech256: lambda dataset: [name.rsplit(".", 1)[1] for name in dataset.categories],
+        datasets.CIFAR10: lambda dataset: dataset.classes,
+        datasets.CIFAR100: lambda dataset: dataset.classes,
+        datasets.FashionMNIST: lambda dataset: dataset.classes,
+        datasets.ImageNet: lambda dataset: [", ".join(names) for names in dataset.classes],
+    }.get(type(dataset))
+    return categories_fn(dataset) if categories_fn is not None else None
+
+
+def classification_wrapper(
+    dataset: datasets.VisionDataset, sample: Tuple[PIL.Image.Image, Optional[int]]
+) -> Tuple[PIL.Image.Image, Optional[datapoints.Label]]:
+    image, label = sample
+    if label is not None:
+        label = datapoints.Label(label, categories=get_categories(dataset))
+    return image, label
+
+
+for dataset_type in [
+    datasets.Caltech256,
+    datasets.CIFAR10,
+    datasets.CIFAR100,
+    datasets.ImageNet,
+    datasets.MNIST,
+    datasets.FashionMNIST,
+]:
+    _WRAPPERS[dataset_type] = classification_wrapper
+
+
+def segmentation_wrapper(
+    dataset: datasets.VisionDataset, sample: Tuple[PIL.Image.Image, PIL.Image.Image]
+) -> Tuple[PIL.Image.Image, datapoints.Mask]:
+    image, mask = sample
+    return image, datapoints.Mask(F.to_image_tensor(mask))
+
+
+for dataset_type in [
+    datasets.VOCSegmentation,
+]:
+    _WRAPPERS[dataset_type] = segmentation_wrapper
+
+
+def caltech101_wrapper(
+    dataset: datasets.Caltech101, sample: Tuple[PIL.Image.Image, Any]
+) -> Tuple[PIL.Image.Image, Any]:
+    image, target = sample
+
+    target_type_wrapper_map: Dict[str, Callable] = {
+        "category": lambda label: datapoints.Label(label, categories=dataset.categories),
+        "annotation": datapoints.GenericDatapoint,
```

Comment:
Is this `annotation` a Tensor already, or is it available as a different type (like a numpy array or a list)? In the spirit of leaving the original types untouched, I wonder if we should just not wrap anything that isn't strictly needed?

Reply:
As discussed offline, `"annotation"` is a numpy array that holds the coordinates of the polygon that represents the outline of the object. Caltech101 is an old datasets, so there only is a single contiguous instance and thus we can represent that with a single array.

We also discussed offline that this is a good use case to avoid the temptation to guess and just fail. This use case is currently not supported by transforms v1 and we aren't using this dataset in our references either. So we can just fail loudly if we encounter this and as the user to create an issue if they need this.